### PR TITLE
Key Verification Support

### PR DESCRIPTION
**Pull request by mollyling**
Two major changes:
- Separates Proxying from Creating a PeerConnection;
- Adds a messaging API on PeerConnection over the control channel to go alongside heartbeats.

---

This change is [Reviewable](https://reviewable.io/reviews/uproxy/uproxy-lib/418)

### CODE REVIEW

**Comment by trevj** [comment]
So, the problem with this is that you're adding features - sending messages - to `PeerConnection` which are already covered by `DataChannel`. The boilerplate implementations of both `sendMessage` and `registerMessageHandler` to both `bridge.ts` and `churn.ts` are a major smell.

Why don't you just create another datachannel and send the verification messages over that? You may not need to change `PeerConnection` or `SocksToRtc` interface at all (I certainly wouldn't object to improving those interfaces but don't go there unless you really need to since it'll require non-trivial changes to uProxy).

**Comment by mollyling** [comment]
The `PeerConnection` already has a control `DataChannel`, and creating new channels on PeerConnections is a mess when the rest of uProxy's designed for them all to be proxy connections  I tried a new-datacannel-dispatcher approach before, for chat, and that got as much bounce back as this.  What do you want here?

**Comment by trevj** [comment]
Ah, I remember that now - `RtcToNet` assumes all new channels on the peerconnection are for proxying. But you could change that, and have it ignore channels that weren't created by `SocksToRtc`:
 * https://github.com/uProxy/uproxy-lib/blob/master/src/rtc-to-net/rtc-to-net.ts#L219

You'll need to examine how `Pool` (currently used by `SocksToRtc`) names channels (basically: `p` + number):
 * https://github.com/uProxy/uproxy-lib/blob/master/src/pool/pool.ts#L92

I think you might be able to avoid making any changes to `PeerConnection`.

**Comment by trevj** [comment]
Bah, `Pool` is a real mess isn't it. After discussing in person, seems like a `getControlChannel`-type method might be the most reasonable/expedient approach for now. Sorry...I agree it's too hard to make any change right now.

**Comment by mollyling** [comment]
Thanks :smiley_cat:

File: src/bridge/bridge.ts [hunk 1]
```diff
@@ -197,7 +198,7 @@ export class BridgingPeerConnection implements peerconnection.PeerConnection<
 
   // Negotiated, actual, provider type.
   private providerType_: ProviderType;
-
+  private queuedHandlers_ :{[name:string]:(name:string, msg:any)=>void} = {};
```
Comment (trevj) [inline src/bridge/bridge.ts]: It doesn't look like this is used.

File: src/webrtc/peerconnection.ts [hunk 2]
```diff
@@ -81,6 +84,9 @@ var HEARTBEAT_INTERVAL_MS_ = 5000;
 // Message which is sent for heartbeats.
 var HEARTBEAT_MESSAGE_ = 'heartbeat';
 
+// Custom message.
```
Comment (trevj) [inline src/webrtc/peerconnection.ts]: Can you think of a more useful comment?

File: src/webrtc/peerconnection.ts [hunk 3]
```diff
@@ -183,6 +189,10 @@ export class PeerConnectionClass implements PeerConnection<signals.Message> {
     this.fulfillClosed_ = F;
   });
 
+  private controlChannel_ :datachannel.ControlChannel = null;
```
Comment (trevj) [inline src/webrtc/peerconnection.ts]: Suggest you instead define a `Promise` that resolves with the control channel. There's a bunch of other such promises here, e.g. this `void`-resolving example:
```typescript
  private fulfillConnected_ :() => void;
  private rejectConnected_ :(e:Error) => void;
  public onceConnected = new Promise<void>((F,R) => {
    this.fulfillConnected_ = F;
    this.rejectConnected_ = R;
  });
```

File: src/webrtc/peerconnection.ts [hunk 4]
```diff
@@ -629,6 +656,96 @@ export class PeerConnectionClass implements PeerConnection<signals.Message> {
   }
 }  // class PeerConnectionClass
 
+// Takes a DataChannel and provides a ControlChannel API atop of it.
+class ControlChannelWrapper implements datachannel.ControlChannel {
```
Comment (trevj) [inline src/webrtc/peerconnection.ts]: This seems like a lot of machinery.

As I see it, `getControlChannel` is a hack to work around how hard it currently is, thanks to `Pool`, to create non-proxying datachannels. Fine, but I'd much rather see further effort go into removing `Pool` and, in the meantime, we can simply mark `getControlChannel` with:
 * `// WARNING: use this channel at your peril, it's used for heartbeat and lifecycle purposes`
 * `// TODO: remove this`

(the warning might be overkill: I can see how the `ControlChannel` interface helps, so long as it's a subset of `DataChannel`)

File: src/webrtc/peerconnection.ts [hunk 5]
```diff
@@ -638,3 +755,15 @@ export function createPeerConnection(
   // closed.
   return new PeerConnectionClass(freedomRtcPc, debugPcName);
 }
+
+// Given a PeerConnection and way for it to send messages, start
+// negotiating a PeerConnection.  Note that the caller is still
+// responsible for sending signals to the PeerConnection (e.g.,
+// invoking handleSignalMessage).
+export function setupPeerConnection(
```
Comment (trevj) [inline src/webrtc/peerconnection.ts]: Do all the sample apps continue to work following this change?

File: src/webrtc/peerconnection.ts (hunk 4 above)
Comment (mollyling) [inline src/webrtc/peerconnection.ts]: I'll add that.  While I'm at it, about half of handleMessage_ is debug-logging-related.  Want me to take that out to shorten this?

**Pull request closed without merging**